### PR TITLE
Refactored PowerBi Connector

### LINKED ISSUE #4947: PowerBI - UI doesn't work due to auth
Powerbi needs to get authenticated on click of a link provided by the powerbi package
image

UI won't work unless the user clicks on the link, paste's the redirect uri.

In order to make this work, we will need to optimize the auth.

### PR DESCRIPTION
**Pull request by OnkarVO7**
### Describe your changes :

Fix #4947

#
### Type of change :

- [x] Bug fix
- [x] Improvement

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers

@open-metadata/ingestion

### CODE REVIEW

File: ingestion/examples/workflows/powerbi.yaml [hunk 1]
```diff
@@ -5,15 +5,19 @@ source:
...
-      - https://analysis.windows.net/powerbi/api/App.Read.All
-      credentials: path
+        - https://analysis.windows.net/powerbi/api/.default
       type: PowerBI
   sourceConfig:
-    config: {}
+    config:
+      chartFilterPattern:
+        includes:
+          - Total Revenue
```
Comment (harshach) [inline ingestion/examples/workflows/powerbi.yaml:17]: isn't this an array? also this should be a regex. Putting a regex pattern would be a better example

Comment (OnkarVO7) [reply ingestion/examples/workflows/powerbi.yaml:17]: yes YAML takes array like these below. I'll add more elements in the array and add a regex for better example

```
chartFilterPattern:
  includes:
    - Total Revenue
    - Total Cost
    - Total Sales
```

File: ingestion/src/metadata/utils/powerbi_client.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+import msal
+
+from metadata.ingestion.ometa.client import REST, ClientConfig
+from metadata.utils.logger import ingestion_logger
+
+logger = ingestion_logger()
+
+
+class PowerBiApiClient:
```
[the remark this answers is not in the document]
Comment (OnkarVO7) [reply ingestion/src/metadata/utils/powerbi_client.py:24]: Yes we can add APIs to retrieve Datasets, Workspaces etc.

[the remark this answers is not in the document]
Comment (pmbrull) [reply ingestion/src/metadata/utils/powerbi_client.py:24]: At least datasets we should, as I understand that they could be ingested as tables. What are workspaces?

[the remark this answers is not in the document]
Comment (OnkarVO7) [reply ingestion/src/metadata/utils/powerbi_client.py:24]: Workspaces contain all these Datasets, dashboards etc and powerbi can have multiple workspaces

[the remark this answers is not in the document]
Comment (pmbrull) [reply ingestion/src/metadata/utils/powerbi_client.py:24]: mmmh oki, IMO if we are able to get all datasets and dashboards we should be good for now

File: ingestion/src/metadata/ingestion/source/dashboard/dashboard_source.py [hunk 3]
```diff
@@ -80,18 +81,17 @@ def __init__(
...
             self.config.sourceConfig.config
         )
-
         self.connection = get_connection(self.service_connection)
         self.client = self.connection.client
         self.service = self.metadata.get_service_or_create(
             entity=DashboardService, config=config
         )
-        self.status = SourceStatus()
+        self.status = SQLSourceStatus()
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/dashboard/dashboard_source.py:89]: Maybe not the best status to use here hehe Opened a ticket to track this https://github.com/open-metadata/OpenMetadata/issues/5247, let's move on for now

File: ingestion/src/metadata/ingestion/source/dashboard/dashboard_source.py [hunk 4]
```diff
@@ -80,18 +81,17 @@ def __init__(
...
         self.metadata_client = OpenMetadata(self.metadata_config)
 
     def next_record(self) -> Iterable[Entity]:
         yield from self.process_dashboards()
-        yield from self.process_charts()
+        yield from self.process_charts() or []
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/dashboard/dashboard_source.py:94]: cool

File: ingestion/src/metadata/ingestion/source/dashboard/powerbi.py [hunk 5]
```diff
@@ -89,70 +79,112 @@ def get_dashboards_list(self) -> Optional[List[dict]]:
...
+                            self.metadata,
+                            entity_type=LineageDashboard,
+                            service_name=self.config.serviceName,
+                            dashboard_name=dashboard_details["id"],
+                        )
+                        to_entity = self.metadata.get_by_name(
+                            entity=LineageDashboard,
+                            fqn=to_fqn,
+                        )
+                        if from_entity and to_entity:
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/dashboard/powerbi.py:148]: nice and clean!

File: ingestion/src/metadata/utils/connections.py [hunk 6]
```diff
@@ -518,3 +522,20 @@ def _(connection: TableauClient) -> None:
...
+def _(connection: PowerBIConnection, verbose: bool = False):
+    from metadata.utils.powerbi_client import PowerBiApiClient
+
+    return PowerBiClient(PowerBiApiClient(connection))
+
+
+@test_connection.register
+def _(connection: PowerBiClient) -> None:
+    try:
+        connection.client.fetch_dashboards()
```
Comment (pmbrull) [inline ingestion/src/metadata/utils/connections.py:537]: is there any other method we can test here that is a bit more lightweight?

Comment (OnkarVO7) [reply ingestion/src/metadata/utils/connections.py:537]: Let me see if there is any other lightweight API we can use

File: ingestion/src/metadata/utils/powerbi_client.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+"""
+import traceback
+from typing import Tuple
+
+import msal
+
+from metadata.ingestion.ometa.client import REST, ClientConfig
+from metadata.utils.logger import ingestion_logger
+
+logger = ingestion_logger()
```
Comment (ulixius9) [inline ingestion/src/metadata/utils/powerbi_client.py:22]: we should use `utils_logger` here instead of `ingestion_logger`

Comment (OnkarVO7) [reply ingestion/src/metadata/utils/powerbi_client.py:22]: ok

File: ingestion/src/metadata/utils/connections.py (hunk 6 above)
Comment (OnkarVO7) [reply ingestion/src/metadata/utils/connections.py:537]: I checked other APIs but they are also similar to this one. These are the list of [supported APIs](https://docs.microsoft.com/en-us/power-bi/enterprise/read-only-apis-service-principal-authentication)

Comment (pmbrull) [reply ingestion/src/metadata/utils/connections.py:537]: oh I see, then let's keep this one for now. Thanks

**Pull request merged**